jsparagus/parse_table.py

```diff
@@ -1314,6 +1322,138 @@ def remove_all_unreachable_state(self, verbose: bool, progress: bool) -> None:
+                            break
+                        assert len(next_aps) == 1
+                        iter_aps = next_aps[0]
+                        next_is_action = self.states[iter_aps.state].epsilon != []
+                        has_replay = iter_aps.replay != []
+                    # print("End at {}:\n{}".format(len(iter_aps.history),
+                    #                               iter_aps.string(name="\titer_aps")))
+                    replay_list = list(map(lambda e: e.src, iter_aps.shift))
+                    assert len(replay_list) >= 2
+                    replay_term = Replay(replay_list[1:])
+                    states_by_replay_term[replay_term].append(replay_list[0])
+
+                # Create FilterStates actions.
+                filter_by_replay_term = {
+                    replay_term: FilterStates(states)
+                    for replay_term, states in states_by_replay_term.items()
+                }
+
+                # Convert the Reduce action to an Unwind action.
+                reduce_term, _ = next(iter(s.epsilon))
+                if isinstance(reduce_term, Reduce):
+                    unwind_term: Action = reduce_term.unwind
+                else:
+                    assert isinstance(reduce_term, Seq)
+                    assert isinstance(reduce_term.actions[-1], Reduce)
+                    unwind_term = Seq(list(reduce_term.actions[:-1]) + [reduce_term.actions[-1].unwind])
+
+                # Remove the old Reduce edge if still present.
+                # print("Before:\n{}".format(reduce_state.stable_str(self.states)))
+                self.remove_edge(reduce_state, reduce_term, maybe_unreachable_set)
+
+                # Add Unwind action.
+                # print("After:\n")
+                locations = OrderedFrozenSet(reduce_state.locations)
+                delayed: OrderedFrozenSet[DelayedAction] = OrderedFrozenSet(filter_by_replay_term.items())
+                is_new, filter_state = self.new_state(locations, delayed)
+                self.add_edge(reduce_state, unwind_term, filter_state.index)
+                if not is_new:
+                    for replay_term, filter_term in filter_by_replay_term.items():
+                        assert filter_term in filter_state
+                        replay_state = self.states[filter_state[filter_term]]
+                        assert replay_term in replay_state
+                        # print(replay_state.stable_str(self.states))
+                    # print(filter_state.stable_str(self.states))
+                    # print(reduce_state.stable_str(self.states))
+                    continue
+
+                for replay_term, filter_term in filter_by_replay_term.items():
+                    dest_idx = replay_term.replay_steps[-1]
+                    dest = self.states[dest_idx]
+
+                    # Add FilterStates action from the filter_state to the replay_state.
+                    locations = OrderedFrozenSet(dest.locations)
```

Comment:
Same comment about `dest.locations`.

Reply:
Er, "same comment" refers to the next comment, on line 1418, "`reduce_state.locations` is already an `OrderedFrozenSet`, so calling `OrderedFrozenSet` on it makes an unnecessary copy." For some reason GitHub shows these out of order.